File: src_fortran/sub/resisave.f
c
c
c
c
c
c
      subroutine RESISAVE(nrp,nrpdeep,rp,nobs,i,k1,mll)
c
c     saves the residual according to the ray-type
c
      implicit none
      integer nrp,nrpdeep,nobs,i,k1,mll
      include '../inc/vel_com.inc'
c
      real xxx,yyy,xhyp,yhyp,xstn,ystn,azi,dist
      integer iazi
      real rp(3,inrpmax)
c
      if(rp(3,nrpdeep).eq.rp(3,nrpdeep+1))then ! ray is horizontal : refracted !
         avrefrres=avrefrres+    res(nobs,i)       ! refracted
         abrefrres=abrefrres+ABS(res(nobs,i))      ! refracted
         nrrefrres=nrrefrres+1
c
         if(lmax.eq.10)then ! refractor is Moho...
            if(irfrout.eq.1)then
c
c              write (x,y) of point A, where ray enters MOHO,
c              of point B, where ray leaves MOHO and of point between A and B
c
               xxx=rp(1,nrpdeep)
               if(icoordsystem.eq.2) xxx=-xxx
               yyy=rp(2,nrpdeep)
               write(78,'(2x,''RFR'',3f10.3,''        .1'')')
     &                    xxx,yyy,res(nobs,i)
               xxx=(rp(1,nrpdeep)+rp(1,nrpdeep+1))/2.
               if(icoordsystem.eq.2) xxx=-xxx
               yyy=(rp(2,nrpdeep)+rp(2,nrpdeep+1))/2.
               write(78,'(2x,''RFR'',3f10.3,''        .1'')')
     &                    xxx,yyy,res(nobs,i)
               xxx=rp(1,nrpdeep+1)
               if(icoordsystem.eq.2) xxx=-xxx
               yyy=rp(2,nrpdeep+1)
               write(78,'(2x,''RFR'',3f10.3,''        .1'')')
     &                    xxx,yyy,res(nobs,i)
             endif
         endif
c
      else
         if(mll.eq.0)then
            avotheres=avotheres+    res(nobs,i)    ! straight or direct
            abotheres=abotheres+ABS(res(nobs,i))   ! straight or direct
            nrotheres=nrotheres+1
         else
            avreflres=avreflres+    res(nobs,i)    ! reflected
            abreflres=abreflres+ABS(res(nobs,i))   ! reflected
            nrreflres=nrreflres+1
c
            if(irflout.eq.1)then
              if(icoordsystem.eq.2)then
                 xxx=-rp(1,nrpdeep)
              else
                 xxx=rp(1,nrpdeep)
              endif
              write(77,'(2x,''RFL'',3f10.4,''        .1'')')
     &                   -rp(1,nrpdeep),rp(2,nrpdeep),res(nobs,i)
            endif
         endif
      endif
c
c     save residual for station# k1 according to the azimuth:
c
      xhyp=rp(1,1)
      yhyp=rp(2,1)
      xstn=rp(1,nrp)
      ystn=rp(2,nrp)
c        Azimuth (stn --> hypoc) = 57.296*ATAN2(xhyp-xstn,yhyp-ystn)
      azi=57.296*ATAN2(xhyp-xstn,yhyp-ystn)
      if(azi.lt.0) azi=azi+360.
      iazi=0
      if(azi.ge.  0.0 .and. azi.lt. 90.0) iazi=1
      if(azi.ge. 90.0 .and. azi.lt.180.0) iazi=2
      if(azi.ge.180.0 .and. azi.lt.270.0) iazi=3
      if(azi.ge.270.0 .and. azi.le.360.0) iazi=4
      stnazires(k1,2*iazi-1)=stnazires(k1,2*iazi-1)+res(nobs,i)
      stnazires(k1,2*iazi)=stnazires(k1,2*iazi)+1.
c
c     type residuals as a function of focus-receiver distance:
      if(iresout.eq.1)then
         dist=(  (rp(1,1)-rp(1,nrp))**2
     &          +(rp(2,1)-rp(2,nrp))**2
     &          +(rp(3,1)-rp(3,nrp))**2  )
         dist=sqrt(dist)
         write(79,'(1x,f6.2,2x,f7.3)') dist,res(nobs,i)
      endif
c
      RETURN
      end ! of subr. resisave

File: src/sub/resisave.c
#include <math.h>
#include <stdio.h>
#include "../include/globals.h"

extern double avrefrres;
extern double abrefrres;
extern int nrrefrres;
extern double avotheres;
extern double abotheres;
extern int nrotheres;
extern double avreflres;
extern double abreflres;
extern int nrreflres;
extern int lmax;
extern int irfrout;
extern int irflout;
extern int iresout;
extern int icoordsystem;
extern double res[IST][IEQ];
extern double stnazires[IST][8];

void resisave(int nrp, int nrpdeep, double rp[3][200], int nobs, int i, int k1, int mll) {
    double xxx, yyy, xhyp, yhyp, xstn, ystn, azi, dist;
    int iazi;

    int obs = nobs - 1;
    int ev = i - 1;
    if (obs < 0 || obs >= IST || ev < 0 || ev >= IEQ) {
        return;
    }
    if (k1 < 0 || k1 >= IST) {
        return;
    }
    if (nrp < 2 || nrp > INRPMAX) {
        return;
    }
    if (nrpdeep < 0 || nrpdeep >= nrp - 1) {
        return;
    }

    if (rp[2][nrpdeep] == rp[2][nrpdeep + 1]) {
        avrefrres += res[obs][ev];
        abrefrres += fabs(res[obs][ev]);
        nrrefrres += 1;

        if (lmax == 10 && irfrout == 1) {
            xxx = rp[0][nrpdeep];
            if (icoordsystem == 2) {
                xxx = -xxx;
            }
            yyy = rp[1][nrpdeep];
            fprintf(stdout, "  RFR%10.3f%10.3f%10.3f        .1\n", xxx, yyy, res[obs][ev]);

            xxx = (rp[0][nrpdeep] + rp[0][nrpdeep + 1]) * 0.5f;
            if (icoordsystem == 2) {
                xxx = -xxx;
            }
            yyy = (rp[1][nrpdeep] + rp[1][nrpdeep + 1]) * 0.5f;
            fprintf(stdout, "  RFR%10.3f%10.3f%10.3f        .1\n", xxx, yyy, res[obs][ev]);

            xxx = rp[0][nrpdeep + 1];
            if (icoordsystem == 2) {
                xxx = -xxx;
            }
            yyy = rp[1][nrpdeep + 1];
            fprintf(stdout, "  RFR%10.3f%10.3f%10.3f        .1\n", xxx, yyy, res[obs][ev]);
        }
    } else {
        if (mll == 0) {
            avotheres += res[obs][ev];
            abotheres += fabs(res[obs][ev]);
            nrotheres += 1;
        } else {
            avreflres += res[obs][ev];
            abreflres += fabs(res[obs][ev]);
            nrreflres += 1;

            if (irflout == 1) {
                if (icoordsystem == 2) {
                    xxx = -rp[0][nrpdeep];
                } else {
                    xxx = rp[0][nrpdeep];
                }
                fprintf(stdout, "  RFL%10.4f%10.4f%10.4f        .1\n", xxx, rp[1][nrpdeep], res[obs][ev]);
            }
        }
    }

    xhyp = rp[0][0];
    yhyp = rp[1][0];
    xstn = rp[0][nrp - 1];
    ystn = rp[1][nrp - 1];

    azi = 57.296f * atan2f(xhyp - xstn, yhyp - ystn);
    if (azi < 0.0) {
        azi += 360.0;
    }

    iazi = 0;
    if (azi >= 0.0 && azi < 90.0) iazi = 1;
    if (azi >= 90.0 && azi < 180.0) iazi = 2;
    if (azi >= 180.0 && azi < 270.0) iazi = 3;
    if (azi >= 270.0 && azi <= 360.0) iazi = 4;

    stnazires[k1][2 * iazi - 2] += res[obs][ev];
    stnazires[k1][2 * iazi - 1] += 1.0f;

    if (iresout == 1) {
        dist = (rp[0][0] - rp[0][nrp - 1]) * (rp[0][0] - rp[0][nrp - 1])
             + (rp[1][0] - rp[1][nrp - 1]) * (rp[1][0] - rp[1][nrp - 1])
             + (rp[2][0] - rp[2][nrp - 1]) * (rp[2][0] - rp[2][nrp - 1]);
        dist = sqrtf(dist);
        fprintf(stdout, " %6.2f  %7.3f\n", dist, res[obs][ev]);
    }
}
